### PR TITLE
bpo-47000: Make `io.text_encoding()` respects UTF-8 mode

### PR DESCRIPTION
**Pull request by methane**
https://bugs.python.org/issue47000

### CODE REVIEW

File: Modules/_io/_iomodule.c [hunk 1]
```diff
@@ -479,7 +480,13 @@ _io_text_encoding_impl(PyObject *module, PyObject *encoding, int stacklevel)
                 return NULL;
             }
         }
-        return &_Py_ID(locale);
```
Comment (methane) [inline Modules/_io/_iomodule.c:482]: @ericsnowcurrently We need to incref this until we implement immortal object. Am I correct?

Comment (ericsnowcurrently) [reply Modules/_io/_iomodule.c:482]: Sort of.  Currently all statically allocated global objects have their refcount initialized to a really large value.  So we haven't been bothering with the incref if we know the object is one of those, which it is in this case.  Incref'ing it won't hurt.  However, it is effectively unnecessary.

File: Modules/_io/_iomodule.c [hunk 2]
```diff
@@ -479,7 +480,13 @@ _io_text_encoding_impl(PyObject *module, PyObject *encoding, int stacklevel)
...
+        const PyPreConfig *preconfig = &_PyRuntime.preconfig;
+        if (preconfig->utf8_mode) {
+            encoding = &_Py_STR(utf_8);
```
Comment (ericsnowcurrently) [inline Modules/_io/_iomodule.c:486]: `_Py_DECLARE_STR()` should always precede use of `_Py_STR()`, even if the same string is declared elsewhere (e.g. `anon_string` in Python/pythonrun.c).  That way things don't break when `_Py_DECLARE_STR()` is dropped elsewhere.  It also makes it clear what the actual string is at the place it's used.  `_Py_DECLARE_STR()` is an empty macro so there is no problem from having duplicates, and Tools/scripts/generate_global_objects.py prevents mismatches.  Sorry all that wasn't more clear.

```suggestion
            _Py_DECLARE_STR(utf_8, "utf-8");
            encoding = &_Py_STR(utf_8);
```

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]

File: Modules/_io/_iomodule.c (hunk 1 above)
Comment (methane) [reply Modules/_io/_iomodule.c:482]: I am afraid breaking refleak buildbot.
Would you tell my why this `return &_Py_ID(locale);` don't break refleak test?

Comment (ericsnowcurrently) [reply Modules/_io/_iomodule.c:482]: The refcount of static objects is not directly included in `_Py_RefTotal`, which is used for refleak detection.  (`_Py_RefTotal` will still ensure that refcount operations are balanced though, regardless of the objects involved.)

Comment (methane) [reply Modules/_io/_iomodule.c:482]: How is it achieved?

https://github.com/python/cpython/blob/cd05d0a423d97be69f9de4650f68f89e99ad68d1/Include/object.h#L473-L515

Py_INCREF and Py_DECREF don't check that the object is statically allocated.

Comment (methane) [reply Modules/_io/_iomodule.c:482]: > (_Py_RefTotal will still ensure that refcount operations are balanced though, regardless of the objects involved.)

So skipping INCREF here might cause in-balance of `_Py_RefTotal`. Doesn't it break refleak test?

Comment (ericsnowcurrently) [reply Modules/_io/_iomodule.c:482]: Good point.  It didn't cause a problem before.  I'd have to look into why not.

File: Modules/_io/_iomodule.c (hunk 2 above)
Comment (ericsnowcurrently) [reply Modules/_io/_iomodule.c:486]: FYI, I added a comment for _Py_DECLARE_STR() in gh-32063.

**Pull request merged**